File: java/src/main/java/gherkin/stream/GherkinEvents.java
package gherkin.stream;

import gherkin.AstBuilder;
import gherkin.Parser;
import gherkin.ParserException;
import gherkin.TokenMatcher;
import gherkin.ast.GherkinDocument;
import gherkin.events.AttachmentEvent;
import gherkin.events.Event;
import gherkin.events.GherkinDocumentEvent;
import gherkin.events.PickleEvent;
import gherkin.events.SourceEvent;
import gherkin.pickles.Compiler;
import gherkin.pickles.Pickle;

import java.util.ArrayList;
import java.util.List;

public class GherkinEvents {
    private final Parser<GherkinDocument> parser = new Parser<>(new AstBuilder());
    private final TokenMatcher matcher = new TokenMatcher();
    private final Compiler compiler = new Compiler();

    private final boolean printSource;
    private final boolean printAst;
    private final boolean printPickles;

    public GherkinEvents(boolean printSource, boolean printAst, boolean printPickles) {
        this.printSource = printSource;
        this.printAst = printAst;
        this.printPickles = printPickles;
    }

    public Iterable<? extends Event> iterable(SourceEvent sourceEvent) {
        List<Event> events = new ArrayList<>();

        try {
            GherkinDocument gherkinDocument = parser.parse(sourceEvent.data, matcher);

            if (printSource) {
                events.add(sourceEvent);
            }
            if (printAst) {
                events.add(new GherkinDocumentEvent(sourceEvent.uri, gherkinDocument));
            }
            if (printPickles) {
                List<Pickle> pickles = compiler.compile(gherkinDocument);
                for (Pickle pickle : pickles) {
                    events.add(new PickleEvent(sourceEvent.uri, pickle));
                }
            }
        } catch (ParserException.CompositeParserException e) {
            for (ParserException error : e.errors) {
                addErrorAttachment(events, error, sourceEvent.uri);
            }
        } catch (ParserException e) {
            addErrorAttachment(events, e, sourceEvent.uri);
        }
        return events;
    }

    private void addErrorAttachment(List<Event> events, ParserException e, String uri) {
        events.add(new AttachmentEvent(
                new AttachmentEvent.SourceRef(
                        uri,
                        new AttachmentEvent.Location(
                                e.location.getLine(),
                                e.location.getColumn()
                        )
                ),
                e.getMessage()
        ));

    }
}

File: python/gherkin/stream/gherkin_events.py
from gherkin.parser import Parser
from gherkin.pickles.compiler import compile
from gherkin.errors import ParserError, CompositeParserException

class GherkinEvents:
    def __init__(self, options):
        self.options = options
        self.parser = Parser()

    def enum(self, source_event):
        uri = source_event['uri']
        source = source_event['data']

        events = []

        try:
            gherkin_document = self.parser.parse(source)

            if (self.options.print_source):
                events.append(source_event)

            if (self.options.print_ast):
                events.append({
                    'type': 'gherkin-document',
                    'uri': uri,
                    'document': gherkin_document
                })

            if (self.options.print_pickles):
                pickles = compile(gherkin_document)
                for pickle in pickles:
                    events.append({
                        'type': 'pickle',
                        'uri': uri,
                        'pickle': pickle
                    })
        except CompositeParserException as e:
            yield_errors(y, e.errors, uri)
        except ParserError as e:
            yield_errors(output, [e], uri)

        return events
